### PR TITLE
Fix GRPC API Validation Flaws

### PR DESCRIPTION
**Pull request by freddy61025**
## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-1925

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: inventory/main/src/test/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcServiceTest.java [hunk 1]
```diff
@@ -102,6 +105,48 @@ void testListMonitoringSystemByLocationId(){
         verify(mockService).findByMonitoringLocationIdAndTenantId(locationId, tenantId);
     }
 
+    @Test
+    void testGgtMonitoringSystemByQuery(){
```
Comment (bouff) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcServiceTest.java:109]: Nit: typo in method name here and line 126

File: events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+public abstract class AbstractGrpcUnitTest {
+
+    protected TenantLookup tenantLookup = new GrpcTenantLookupImpl();
+    protected String serverName;
+    protected Server server;
+
+    protected EventServerInterceptor spyInterceptor;
+
+    protected final String tenantId = "test-tenant";
```
Comment (bouff) [inline events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java:60]: Nit: could be constants (here and line below)

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java [hunk 3]
```diff
@@ -101,6 +109,17 @@ public MinionCertificateManagerImpl(@Value("${manager.mtls.certificate}") File c
 
         LOG.info("CA EXISTS: {}, CA PATH {}, CA CAN READ {}", caCertFile.exists(), caCertFile.getAbsolutePath(), caCertFile.canRead());
         this.serialNumberRepository = serialNumberRepository;
+        this.inventoryClient = Objects.requireNonNull(inventoryClient);
+    }
+
+    boolean isLocationExist(long locationId, String tenantId) {
```
Comment (bouff) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java:115]: Nit: can be private?

File: inventory/main/src/test/java/org/opennms/horizon/inventory/service/TagServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        tag2.setMonitorPolicyIds(new ArrayList<>(List.of(testPolicyId)));
+        tag2.setTenantId(TEST_TENANT_ID);
+        when(mockAlertClient.getPolicyById(testPolicyId, TEST_TENANT_ID)).thenReturn(MonitorPolicyProto.newBuilder().build());
+        when(mockTagRepository.findByTenantIdAndId(eq(TEST_TENANT_ID), any())).thenAnswer((args) -> {
+            long id = args.getArgument(1);
+            if (id == tag1.getId()) {
+                return Optional.of(tag1);
+            } else if (id == tag2.getId()) {
+                return Optional.of(tag2);
+
```
Comment (bouff) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/service/TagServiceTest.java:248]: Nit: can remove newline here

**Review by bouff — COMMENTED** [review]
Overall LGTM.  Only a couple minor nits.

File: inventory/main/src/test/java/org/opennms/horizon/inventory/component/AlertClientTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+class AlertClientTest {
+    @Rule
+    public final static GrpcCleanupRule grpcCleanUp = new GrpcCleanupRule();
```
Comment (vserrago) [inline inventory/main/src/test/java/org/opennms/horizon/inventory/component/AlertClientTest.java:58]: As mentioned previously in a previous location, this rule is mis-specified due to using Junit5 annotations and a static field.

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java [hunk 6]
```diff
@@ -76,15 +82,17 @@ public class MinionCertificateManagerImpl extends MinionCertificateManagerGrpc.M
...
     @Autowired
     public MinionCertificateManagerImpl(@Value("${manager.mtls.certificate}") File certificate,
-        @Value("${manager.mtls.privateKey}") File privateKey, @Autowired SerialNumberRepository serialNumberRepository) throws IOException, InterruptedException {
-        this(certificate, privateKey, new PKCS12Generator(), serialNumberRepository);
+        @Value("${manager.mtls.privateKey}") File privateKey, @Autowired SerialNumberRepository serialNumberRepository, @Autowired InventoryClient inventoryClient) throws IOException, InterruptedException {
+        this(certificate, privateKey, new PKCS12Generator(), serialNumberRepository, inventoryClient);
     }
 
-    MinionCertificateManagerImpl(File certificate, File key, PKCS12Generator pkcs8Generator, SerialNumberRepository serialNumberRepository) throws IOException, InterruptedException {
+    MinionCertificateManagerImpl(File certificate, File key, PKCS12Generator pkcs8Generator, SerialNumberRepository serialNumberRepository, InventoryClient inventoryClient) throws IOException, InterruptedException {
```
Comment (vserrago) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java:95]: Nit: Line is excessively long and would benefit from being broken into multiple lines.

```suggestion
    MinionCertificateManagerImpl(
        File certificate,
        File key,
        PKCS12Generator pkcs8Generator,
        SerialNumberRepository serialNumberRepository,
        InventoryClient inventoryClient
    ) throws IOException, InterruptedException {
```

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java [hunk 7]
```diff
@@ -76,15 +82,17 @@ public class MinionCertificateManagerImpl extends MinionCertificateManagerGrpc.M
...
 
+    private final InventoryClient inventoryClient;
+
     private CommandExecutor commandExecutor = new CommandExecutor();
 
     @Autowired
     public MinionCertificateManagerImpl(@Value("${manager.mtls.certificate}") File certificate,
-        @Value("${manager.mtls.privateKey}") File privateKey, @Autowired SerialNumberRepository serialNumberRepository) throws IOException, InterruptedException {
-        this(certificate, privateKey, new PKCS12Generator(), serialNumberRepository);
+        @Value("${manager.mtls.privateKey}") File privateKey, @Autowired SerialNumberRepository serialNumberRepository, @Autowired InventoryClient inventoryClient) throws IOException, InterruptedException {
```
Comment (vserrago) [inline minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java:91]: Nit: Line is excessively long and would benefit from being broken into multiple lines.

```suggestion
        @Value("${manager.mtls.privateKey}") File privateKey,
        @Autowired SerialNumberRepository serialNumberRepository,
        @Autowired InventoryClient inventoryClient
    ) throws IOException, InterruptedException {
```

**Review by vserrago — CHANGES_REQUESTED** [review]
Looks good overall, but there are a few items to address. Any nits are optional.

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java (hunk 7 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java]: done

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java (hunk 6 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java]: done

File: inventory/main/src/test/java/org/opennms/horizon/inventory/component/AlertClientTest.java (hunk 5 above)
Comment (freddy61025) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/component/AlertClientTest.java]: done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/active/AzureActiveDiscoveryService.java [hunk 8]
```diff
@@ -95,6 +98,12 @@ private void validateDiscovery(String tenantId, AzureActiveDiscoveryCreateDTO re
         } catch (Exception e) {
             throw new InventoryRuntimeException("Failed to login with azure credentials", e);
         }
+
+        var location = monitoringLocationService.findByLocationIdAndTenantId(Long.parseLong(request.getLocationId()), tenantId);
+        if (location.isEmpty()) {
+            throw new LocationNotFoundException("Location not found.");
+        }
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/discovery/active/AzureActiveDiscoveryService.java]: done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/TagService.java [hunk 9]
```diff
@@ -510,4 +524,14 @@ public void insertOrUpdateTags(TagOperationList list) {
             }
         });
     }
+
+    private boolean isPolicyExist(long policyId, String tenantId) {
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/TagService.java]: done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcService.java [hunk 10]
```diff
@@ -68,6 +68,14 @@ public void listMonitoringSystemByLocationId(Int64Value locationId, StreamObserv
...
+
+        if (list.isEmpty()) {
+            responseObserver.onError(StatusProto.toStatusRuntimeException(createStatusNotExist(locationId.getValue())));
+        } else {
+            responseObserver.onNext(MonitoringSystemList.newBuilder().addAllSystems(list).build());
+            responseObserver.onCompleted();
+        }
+
         responseObserver.onNext(MonitoringSystemList.newBuilder().addAllSystems(list).build());
         responseObserver.onCompleted();
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcService.java]: nice catch

File: events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+import io.grpc.BindableService;
+import io.grpc.Metadata;
+import io.grpc.Server;
+import io.grpc.ServerInterceptors;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.testing.GrpcCleanupRule;
+
+public abstract class AbstractGrpcUnitTest {
+
+    protected TenantLookup tenantLookup = new GrpcTenantLookupImpl();
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java]: done

File: events/grpc/src/test/java/org/opennms/horizon/events/grpc/client/InventoryClientTest.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+import static org.mockito.AdditionalAnswers.delegatesTo;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.reset;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+public class InventoryClientTest {
+    @Rule
+    public static final GrpcCleanupRule grpcCleanUp = new GrpcCleanupRule();
```
[the remark this answers is not in the document]
Comment (freddy61025) [reply events/grpc/src/test/java/org/opennms/horizon/events/grpc/client/InventoryClientTest.java]: done

File: inventory/main/src/test/java/org/opennms/horizon/inventory/service/TagServiceTest.java (hunk 4 above)
Comment (freddy61025) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/service/TagServiceTest.java]: done

File: minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java (hunk 3 above)
Comment (freddy61025) [reply minion-certificate-manager/main/src/main/java/org/opennms/horizon/minioncertmanager/grpc/MinionCertificateManagerImpl.java]: done

File: events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java (hunk 2 above)
Comment (freddy61025) [reply events/grpc/src/test/java/org/opennms/horizon/events/grpc/service/AbstractGrpcUnitTest.java:58]: done

File: inventory/main/src/test/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcServiceTest.java (hunk 1 above)
Comment (freddy61025) [reply inventory/main/src/test/java/org/opennms/horizon/inventory/grpc/MonitoringSystemGrpcServiceTest.java]: done

**Review by vserrago — APPROVED** [review]

**Review by vserrago — APPROVED** [review]

**Pull request merged**